airflow/logging/airflow_task_handler.py

```diff
@@ -0,0 +1,244 @@
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+import logging
+import os
+import warnings
+
+from airflow import configuration as conf
+from airflow.configuration import AirflowConfigException
+from airflow.utils import logging as logging_utils
+from airflow.utils.file import mkdirs
+from airflow.utils.state import State
+
+
+class AirflowTaskHandler(logging.Handler):
+    """
+    AirflowTaskHandler is the python log handler that handles and reads
+    task instance logs. It creates and delegates log handling
+    to `logging.FileHandler` after receiving task instance context.
+    Upon close, it uploads the log to remote storage like S3/GCS.
+    """
+
+    def __init__(self):
+        super(AirflowTaskHandler, self).__init__()
+        # Setup dummy class variables before receiving task instance to
+        # prevent AttributeError when system shuts down.
+        self.handler = logging.NullHandler()
+        self.filename = ''
+        self.closed = False
+
+    def set_context(self, task_instance):
+        '''
+        Provide task_instance context to airflow task handler.
+        :param task_instance: task instance object
+        '''
+        self.filename = self._init_file(task_instance)
+        self.handler = logging.FileHandler(self.filename)
+        self.handler.setFormatter(self.formatter)
+        self.handler.setLevel(self.level)
+
+    def emit(self, record):
+        self.handler.emit(record)
+
+    def flush(self):
+        self.handler.flush()
+
+    def close(self):
+        '''
+        Upload to remote storage (S3 or GCS) in addition to closing the file.
+        '''
+        # When application exit, system shuts down all handlers by calling
+        # close method. Here we check if logger is already closed to prevent
+        # uploading the log to remote storage multiple times when `logging.shutdown`
+        # is called.
+        if self.closed:
+            return
+
+        self.handler.close()
+
+        # Upload to remote storage
+        remote_base = conf.get('core', 'REMOTE_BASE_LOG_FOLDER')
+
+        # deprecated as of March 2016
+        if not remote_base and conf.get('core', 'S3_LOG_FOLDER'):
+            warnings.warn(
+                'The S3_LOG_FOLDER conf key has been replaced by '
+                'REMOTE_BASE_LOG_FOLDER. Your conf still works but please '
+                'update airflow.cfg to ensure future compatibility.',
+                DeprecationWarning)
+            remote_base = conf.get('core', 'S3_LOG_FOLDER')
+
+        log_base = os.path.expanduser(conf.get('core', 'BASE_LOG_FOLDER'))
+        relative_path = self.filename
+        filename = os.path.join(log_base, relative_path)
+
+        # TODO: Ideally we want to have S3AirflowTaskHandler and
```

Comment:
So do you think we should do it in this PR or a separate one? The refactoring might not be hard but it will be cleaner to do in a separate PR.